sdk/trace/src/main/scala/org/typelevel/otel4s/sdk/trace/IdGenerator.scala

```diff
@@ -0,0 +1,81 @@
+/*
+ * Copyright 2023 Typelevel
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.typelevel.otel4s.sdk.trace
+
+import cats.Monad
+import cats.effect.std.Random
+import cats.syntax.flatMap._
+import cats.syntax.functor._
+import cats.syntax.monad._
+import org.typelevel.otel4s.trace.SpanContext
+import scodec.bits.ByteVector
+
+/** Generates new span and trace ids.
+  *
+  * @see
+  *   [[https://opentelemetry.io/docs/specs/otel/trace/sdk/#id-generators]]
+  *
+  * @tparam F
+  *   the higher-kinded type of a polymorphic effect
+  */
+trait IdGenerator[F[_]] {
+
+  /** Generates a valid span id.
+    */
+  def generateSpanId: F[ByteVector]
+
+  /** Generates a valid trace id.
+    */
+  def generateTraceId: F[ByteVector]
+
+  /** Whether it's safe to skip the ID validation: we are sure the generated ids
+    * are valid.
+    */
+  private[trace] def canSkipIdValidation: Boolean =
+    false
+}
+
+object IdGenerator {
+
+  private val InvalidId = 0
+
+  /** Creates [[IdGenerator]] that uses [[cats.effect.std.Random]] under the
+    * hood.
+    */
+  def random[F[_]: Monad: Random]: IdGenerator[F] =
+    new RandomIdGenerator[F]
+
+  private final class RandomIdGenerator[F[_]: Monad: Random]
+      extends IdGenerator[F] {
+
+    def generateSpanId: F[ByteVector] =
+      Random[F].nextLong
+        .iterateUntil(_ != InvalidId)
+        .map(SpanContext.SpanId.fromLong)
+
+    def generateTraceId: F[ByteVector] =
+      for {
+        hi <- Random[F].nextLong
+        lo <- Random[F].nextLong.iterateUntil(_ != InvalidId)
+      } yield SpanContext.TraceId.fromLongs(hi, lo)
```

Comment:
What are the rules for trace ids? is it not a valid trace id if `hi != 0` but `lo == 0`?

Reply:
Yes. At least it's correct for Java SDK: https://github.com/open-telemetry/opentelemetry-java/blob/main/sdk/trace/src/main/java/io/opentelemetry/sdk/trace/RandomIdGenerator.java#L32-L38

Reply:
> A valid trace identifier is a 16-byte array with at least one non-zero byte.

Isn't that ID generator missing lots and lots of possible valid IDs? Is that intentional 🤔

Reply:
Aha, there is some explanation here in https://github.com/open-telemetry/opentelemetry-java/pull/3291#discussion_r646688186. I'm not sure if that's still relevant, or relevant to us.

Reply:
I guess we can stick with Java SDK implementation

Reply:
In that case:
```suggestion
    def generateTraceId: F[ByteVector] =
      for {
        hi <- Random[F].nextLong
        lo <- Random[F].nextLongBounded(Long.MaxValue)
      } yield SpanContext.TraceId.fromLongs(hi, lo+1)
```

Reply:
Heh, the execution time of tests went from:
```
org.typelevel.otel4s.sdk.trace.IdGeneratorSuite:
  + generate a valid trace id 0.636s
  + generate a valid span id 0.36s

```
to:
```
org.typelevel.otel4s.sdk.trace.IdGeneratorSuite:
  + generate a valid trace id 9.87s
  + generate a valid span id 10.375s
```

I guess we should stick with `iterateUntil` rather than `nextLongBounded`.

Reply:
😱

Reply:
On the other hand, you rarely need to allocate 1 million random uuids

Reply:
`nextLong`:
```scala
      for {
        r <- f
        out <- Sync[F].delay(r.nextLong())
      } yield out
```

`nextLongBounded`:
```scala
    def nextLongBounded(n: Long): F[Long] = {
      /*
       * Divide n by two until small enough for nextInt. On each
       * iteration (at most 31 of them but usually much less),
       * randomly choose both whether to include high bit in result
       * (offset) and whether to continue with the lower vs upper
       * half (which makes a difference only if odd).
       */
      for {
        _ <- require(n > 0, s"n must be positive, but was $n")
        offset <- Ref[F].of(0L)
        _n <- Ref[F].of(n)
        _ <- Monad[F].whileM_(_n.get.map(_ >= Integer.MAX_VALUE))(
          for {
            bits <- nextIntBounded(2)
            halfn <- _n.get.map(_ >>> 1)
            nextN <- if ((bits & 2) == 0) halfn.pure[F] else _n.get.map(_ - halfn)
            _ <-
              if ((bits & 1) == 0) _n.get.flatMap(n => offset.update(_ + (n - nextN)))
              else Applicative[F].unit
            _ <- _n.set(nextN)
          } yield ()
        )
        finalOffset <- offset.get
        int <- _n.get.flatMap(l => nextIntBounded(l.toInt))
      } yield finalOffset + int
    }
```

Reply:
Gosh, is that really the implementation? I assumed it was delegating to the underlying `Random` instance, not that there are atomics in there 🤦

Reply:
I believe I'm look into the right one :D 

https://github.com/typelevel/cats-effect/blob/9ff20a5ce17e91b0a61addf7f724e498cb13578e/std/shared/src/main/scala/cats/effect/std/Random.scala#L403